grpc/src/main/proto/tree_service.proto

```diff
@@ -0,0 +1,192 @@
+
+service TreeService {
+  rpc getAllReferences (Empty) returns (GetAllReferencesResponse) {}
+  rpc getReferenceByName (GetReferenceByNameRequest) returns (Reference) {}
+  rpc createReference (CreateReferenceRequest) returns (Reference) {}
+  rpc getDefaultBranch (Empty) returns (Reference) {}
+  rpc assignTag (AssignReferenceRequest) returns (Empty) {}
+  rpc deleteTag (DeleteReferenceRequest) returns (Empty) {}
+  rpc assignBranch (AssignReferenceRequest) returns (Empty) {}
+  rpc deleteBranch (DeleteReferenceRequest) returns (Empty) {}
+  rpc getCommitLog (CommitLogRequest) returns (CommitLogResponse) {}
+  rpc getEntries (EntriesRequest) returns (EntriesResponse) {}
+  rpc transplantCommitsIntoBranch (TransplantRequest) returns (Empty) {}
+  rpc mergeRefIntoBranch (MergeRequest) returns (Empty) {}
+  rpc commitMultipleOperations (CommitRequest) returns (Branch) {}
+}
+
+message Reference {
+  oneof type {
+    Branch branch = 1;
+    Tag tag = 2;
+    Hash hash = 3;
+  }
+}
+
+message Branch {
+  string name = 1;
+  string hash = 2;
+}
+
+message Tag {
+  string name = 1;
+  string hash = 2;
+}
+
+message Hash {
+  string hash = 1;
+}
+
+message CreateReferenceRequest {
+  string sourceRefName = 1;
+  Reference reference = 2;
+}
+
+message GetAllReferencesResponse {
+  repeated Reference reference = 1;
+}
+
+message GetReferenceByNameRequest {
+  string namedRef = 1;
+}
+
+message AssignReferenceRequest {
+  string namedRef = 1;
+  string oldHash = 2;
+  oneof assignTo {
+    Branch branch = 3;
+    Tag tag = 4;
+  }
+}
+
+message DeleteReferenceRequest {
+  string namedRef = 1;
+  string hash = 2;
+}
+
+message CommitLogRequest {
+  string namedRef = 1;
+  string startHash = 2;
+  string endHash = 3;
+  int32 maxRecords = 4;
+  string pageToken = 5;
+  string queryExpression = 6;
+}
+
+message CommitLogResponse {
+  repeated CommitMeta operations = 1;
+  bool hasMore = 2;
+  string token = 3;
+}
+
+message CommitMeta {
+  string hash = 1;
+  string committer = 2;
+  string author = 3;
+  string signedOffBy = 4;
+  string message = 5;
+  google.protobuf.Timestamp commitTime = 6;
+  google.protobuf.Timestamp authorTime = 7;
+  map<string, string> properties = 8;
+}
+
+message EntriesRequest {
+  string namedRef = 1;
+  string hashOnRef = 2;
+  int32 maxRecords = 3;
+  string pageToken = 4;
+  string queryExpression = 5;
+}
+
+message EntriesResponse {
+  repeated Entry entries = 1;
+  bool hasMore = 2;
+  string token = 3;
+}
+
+enum ContentsType {
+  UNKNOWN = 0;
+  ICEBERG_TABLE = 1;
+  DELTA_LAKE_TABLE = 2;
+  HIVE_TABLE = 3;
+  HIVE_DATABASE = 4;
+  VIEW = 5;
+}
+
+message Entry {
+  ContentsType type = 1;
+  ContentsKey contentsKey = 2;
+}
+
+message TransplantRequest {
+  string branchName = 1;
+  string hash = 2;
+  string message = 3;
+  string fromRefName = 4;
+  repeated string hashesToTransplant = 5;
+}
+
+message MergeRequest {
+  string fromHash = 1;
+  string toBranch = 2;
+  string expectedHash = 3;
+  string fromRefName = 4;
```

Comment:
nit: inconsistent parameter order... keep related params together (e.g. `fromRefName` and `fromHash`)?

Reply:
I wonder it we could systematically use `Reference` (and sub-types) for arguments that have a "name" and "expected hash"... for example `MergeRequest` could be:
```
message MergeRequest {
  Reference from = 1;
  Branch to = 2;
}
```
WDYT?

Reply:
^ that is not really specific to the gRPC stuff, but more of a general API review for 1.0, I guess :)

Reply:
yeah overall I think we should fix stuff like this first in the REST API

Reply:
Yeah, I mentioned that in #1804 . Let's review APIs under that issue :+1: